File: backend/ble.py
import asyncio
from typing import Callable

from backend.decode import decode_frame_u16, decode_payload
from bleak import BleakClient, BleakScanner
UART_SERVICE_UUID = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"
UART_RX_CHAR_UUID = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E"  # central -> peripheral
UART_TX_CHAR_UUID = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E"  # peripheral -> central


class MagicFrameAssembler:
    """
    Assembles fixed-length frames from a byte stream by realigning on a 2-byte
    little-endian magic header (default 0xBEEF -> b'\\xEF\\xBE').

    If use_sequence=True, expects each BLE notification chunk to begin with a
    2-byte little-endian sequence number, which is stripped before buffering.
    """
    def __init__(self, frame_len: int, magic: int = 0xBEEF, use_sequence: bool = False) -> None:
        self.frame_len = frame_len
        self.use_sequence = use_sequence
        self._buffer = bytearray()
        self._expected_seq = 0

        self._magic = magic
        self._magic_bytes = bytes([magic & 0xFF, (magic >> 8) & 0xFF])  # little-endian

    def add_chunk(self, data: bytes) -> list[bytes]:
        if not data:
            return []

        # Optional 2-byte seq header per notification
        if self.use_sequence:
            if len(data) < 2:
                return []
            seq = data[0] | (data[1] << 8)
            if seq != self._expected_seq:
                # sequence jumped -> drop buffer to avoid mixing partial frames
                self._buffer.clear()
                self._expected_seq = seq
            self._expected_seq = (seq + 1) & 0xFFFF
            data = data[2:]
            if not data:
                return []

        self._buffer.extend(data)
        out: list[bytes] = []

        while True:
            # Find magic start
            i = self._buffer.find(self._magic_bytes)
            if i == -1:
                # Keep last byte in case magic splits across chunks (EF | BE)
                if len(self._buffer) > 1:
                    self._buffer[:] = self._buffer[-1:]
                return out

            # Drop any junk before magic
            if i > 0:
                del self._buffer[:i]

            # Need full frame
            if len(self._buffer) < self.frame_len:
                return out

            frame = bytes(self._buffer[:self.frame_len])
            del self._buffer[:self.frame_len]
            out.append(frame)



async def _find_device_by_name(device_name: str, timeout: float = 10.0):
    def _match_name(d, _):
        return d and d.name == device_name

    return await BleakScanner.find_device_by_filter(_match_name, timeout=timeout)


async def read_one_payload(
    device_name: str,
    payload_len: int,
    timeout: float = 10.0,
    use_sequence: bool = False,
) -> bytes:
    device = await _find_device_by_name(device_name, timeout=timeout)
    if device is None:
        raise RuntimeError(f"BLE device not found: {device_name}")

    assembler = MagicFrameAssembler(payload_len)
    payload_future: asyncio.Future[bytes] = asyncio.get_event_loop().create_future()

    def _handler(_, data: bytearray):
        for payload in assembler.add_chunk(bytes(data)):
            if not payload_future.done():
                payload_future.set_result(payload)

    async with BleakClient(device) as client:
        await client.start_notify(UART_TX_CHAR_UUID, _handler)
        payload = await asyncio.wait_for(payload_future, timeout=timeout)
        await client.stop_notify(UART_TX_CHAR_UUID)
        return payload


async def stream_payloads(
    device_name: str,
    payload_len: int,
    on_payload: Callable[[bytes], None],
    use_sequence: bool = False,
) -> None:
    device = await _find_device_by_name(device_name)
    if device is None:
        raise RuntimeError(f"BLE device not found: {device_name}")

    assembler = MagicFrameAssembler(payload_len)

    def _handler(_, data: bytearray):
        for payload in assembler.add_chunk(bytes(data)):
            on_payload(payload)

    async with BleakClient(device) as client:
        await client.start_notify(UART_TX_CHAR_UUID, _handler)
        try:
            while True:
                await asyncio.sleep(1.0)
        finally:
            await client.stop_notify(UART_TX_CHAR_UUID)


def _print_matrix(matrix) -> None:
    for row in matrix:
        print(", ".join(f"{v:.1f}" for v in row))


def _show_heatmap(matrix, rows: int = 12, cols: int = 8) -> None:
    import matplotlib
    import numpy as np

    if not hasattr(_show_heatmap, "_configured"):
        try:
            matplotlib.use("TkAgg")
        except Exception:
            pass
        _show_heatmap._configured = True

    import matplotlib.pyplot as plt

    vmin = -1.0
    vmax = 3700.0

    arr = np.array(matrix, dtype=float)
    if arr.ndim == 1 and arr.size == rows * cols:
        arr = arr.reshape((rows, cols))
    if arr.ndim != 2:
        return

    if not hasattr(_show_heatmap, "_im"):
        _show_heatmap._fig, _show_heatmap._ax = plt.subplots()
        _show_heatmap._im = _show_heatmap._ax.imshow(
            np.clip(arr, vmin, vmax),
            cmap="inferno_r",
            aspect="auto",
            vmin=vmin,
            vmax=vmax,
        )
        _show_heatmap._ax.set_title("Pressure Heatmap")
        _show_heatmap._fig.colorbar(_show_heatmap._im, ax=_show_heatmap._ax)
        _show_heatmap._texts = []
        plt.ion()
        plt.show(block=False)
    else:
        _show_heatmap._im.set_data(np.clip(arr, vmin, vmax))

    # Overlay raw values as text
    arr = np.clip(arr, vmin, vmax)
    for t in getattr(_show_heatmap, "_texts", []):
        t.remove()
    _show_heatmap._texts = []
    rows, cols = arr.shape
    for r in range(rows):
        for c in range(cols):
            _show_heatmap._texts.append(
                _show_heatmap._ax.text(
                    c,
                    r,
                    f"{arr[r, c]:.1f}",
                    ha="center",
                    va="center",
                    color="white",
                    fontsize=6,
                )
            )

    _show_heatmap._fig.canvas.draw_idle()
    _show_heatmap._fig.canvas.flush_events()
    plt.pause(0.001)


def _parse_args():
    import argparse

    parser = argparse.ArgumentParser(description="BLE UART test receiver")
    parser.add_argument("--name", required=True, help="BLE device name")
    parser.add_argument("--rows", type=int, default=12, help="matrix rows")
    parser.add_argument("--cols", type=int, default=8, help="matrix cols")
    parser.add_argument(
        "--min", dest="min_v", type=float, default=100.0, help="min range"
    )
    parser.add_argument(
        "--max", dest="max_v", type=float, default=100000.0, help="max range"
    )
    parser.add_argument(
        "--seq",
        action="store_true",
        help="expect 2-byte sequence header per chunk",
    )
    parser.add_argument(
        "--once",
        action="store_true",
        help="read one payload then exit",
    )
    return parser.parse_args()


async def _main():
    try:
        from backend.decode import decode_payload
    except ModuleNotFoundError:
        import os
        import sys

        repo_root = os.path.dirname(os.path.dirname(__file__))
        if repo_root not in sys.path:
            sys.path.insert(0, repo_root)
        from backend.decode import decode_payload

    args = _parse_args()
    payload_len = 4 + args.rows * args.cols * 2   # 196

    loop = asyncio.get_running_loop()
    queue: asyncio.Queue[bytes] = asyncio.Queue(maxsize=2)

    def _on_payload(payload: bytes) -> None:
        loop.call_soon_threadsafe(queue.put_nowait, payload)

    stream_task = asyncio.create_task(
        stream_payloads(
            args.name,
            payload_len,
            on_payload=_on_payload,
            use_sequence=args.seq,
        )
    )

    try:
        while True:
            payload = await queue.get()
            frame_id, matrix = decode_frame_u16(payload, min_v=-1.0, max_v=3700.0, rows=12, cols=8)
            print("frame_id:", frame_id)


            import numpy as np

            np.set_printoptions(suppress=True, precision=2)
            print(matrix)
            _show_heatmap(matrix, rows=args.rows, cols=args.cols)
    finally:
        stream_task.cancel()


if __name__ == "__main__":
    asyncio.run(_main())

File: frontend/src/services/ble.js
export const UART_SERVICE_UUID = '6e400001-b5a3-f393-e0a9-e50e24dcca9e';
export const UART_RX_CHAR_UUID = '6e400002-b5a3-f393-e0a9-e50e24dcca9e';
export const UART_TX_CHAR_UUID = '6e400003-b5a3-f393-e0a9-e50e24dcca9e';

const toUint8Array = (payload) => {
  if (payload instanceof Uint8Array) return payload;
  if (payload instanceof ArrayBuffer) return new Uint8Array(payload);
  if (Array.isArray(payload)) return Uint8Array.from(payload);
  return new Uint8Array([]);
};

export class ChunkAssembler {
  constructor(payloadLen, useSequence = false) {
    this.payloadLen = payloadLen;
    this.useSequence = useSequence;
    this.buffer = new Uint8Array(0);
    this.expectedSeq = 0;
  }

  addChunk(data) {
    const chunk = toUint8Array(data);
    if (!chunk.length) return [];

    let offset = 0;
    if (this.useSequence) {
      if (chunk.length < 2) return [];
      const seq = chunk[0] | (chunk[1] << 8);
      if (seq !== this.expectedSeq) {
        this.buffer = new Uint8Array(0);
        this.expectedSeq = seq;
      }
      this.expectedSeq = (seq + 1) & 0xffff;
      offset = 2;
    }

    const remaining = chunk.slice(offset);
    if (!remaining.length) return [];

    const combined = new Uint8Array(this.buffer.length + remaining.length);
    combined.set(this.buffer, 0);
    combined.set(remaining, this.buffer.length);
    this.buffer = combined;

    const complete = [];
    while (this.buffer.length >= this.payloadLen) {
      complete.push(this.buffer.slice(0, this.payloadLen));
      this.buffer = this.buffer.slice(this.payloadLen);
    }

    return complete;
  }
}

export class MagicFrameAssembler {
    constructor(frameLen, magic = 0xbeef) {
      this.frameLen = frameLen;
      this.magicLo = magic & 0xff;
      this.magicHi = (magic >> 8) & 0xff;
      this.buffer = new Uint8Array(0);
    }
  
    addChunk(data) {
      const chunk = toUint8Array(data);
      if (!chunk.length) return [];
  
      // append to buffer
      const combined = new Uint8Array(this.buffer.length + chunk.length);
      combined.set(this.buffer, 0);
      combined.set(chunk, this.buffer.length);
      this.buffer = combined;
  
      const out = [];
  
      while (true) {
        // find magic 0xBEEF (little endian: EF BE)
        let start = -1;
        for (let i = 0; i + 1 < this.buffer.length; i += 1) {
          if (this.buffer[i] === this.magicLo && this.buffer[i + 1] === this.magicHi) {
            start = i;
            break;
          }
        }
  
        if (start === -1) {
          // keep last byte in case magic splits across chunks
          if (this.buffer.length > 1) {
            this.buffer = this.buffer.slice(this.buffer.length - 1);
          }
          return out;
        }
  
        // drop junk before magic
        if (start > 0) {
          this.buffer = this.buffer.slice(start);
        }
  
        // wait for full frame
        if (this.buffer.length < this.frameLen) {
          return out;
        }
  
        out.push(this.buffer.slice(0, this.frameLen));
        this.buffer = this.buffer.slice(this.frameLen);
      }
    }
  }
  

export const connectBleUart = async ({
  name,
  serviceUuid = UART_SERVICE_UUID,
} = {}) => {
  if (!navigator?.bluetooth) {
    throw new Error('Web Bluetooth is not supported in this browser.');
  }

  const filters = [];
  console.log('connectBleUart: name=', name);
  if (name) filters.push({ name });

  const device = await navigator.bluetooth.requestDevice({
    filters: filters.length ? filters : undefined,
    acceptAllDevices: filters.length ? undefined : true,
    optionalServices: [serviceUuid.toLowerCase()],
  });

  if (!device?.gatt) {
    throw new Error('No GATT server available on selected device.');
  }

  const server = await device.gatt.connect();
  const service = await server.getPrimaryService(serviceUuid);
  const txCharacteristic = await service.getCharacteristic(UART_TX_CHAR_UUID);
  const rxCharacteristic = await service.getCharacteristic(UART_RX_CHAR_UUID);

  return { device, server, txCharacteristic, rxCharacteristic };
};

export const disconnectBleDevice = async (device) => {
  if (device?.gatt?.connected) {
    device.gatt.disconnect();
  }
};

export const startUartNotifications = async (
  characteristic,
  { payloadLen, useSequence = false, onPayload }
) => {
  if (!characteristic) {
    throw new Error('Missing UART TX characteristic.');
  }
  if (typeof onPayload !== 'function') {
    throw new Error('onPayload callback is required.');
  }

  //const assembler = new ChunkAssembler(payloadLen, useSequence);
  console.log('startUartNotifications', payloadLen);
  const assembler = new MagicFrameAssembler(payloadLen);
  const handler = (event) => {
    const value = event?.target?.value;
    if (!value) return;
    const bytes = new Uint8Array(
      value.buffer.slice(value.byteOffset, value.byteOffset + value.byteLength)
    );
    for (const payload of assembler.addChunk(bytes)) {
      onPayload(payload);
    }
  };

  characteristic.addEventListener('characteristicvaluechanged', handler);
  await characteristic.startNotifications();

  return async () => {
    characteristic.removeEventListener('characteristicvaluechanged', handler);
    try {
      await characteristic.stopNotifications();
    } catch (e) {
      // Ignore stop errors when device disconnects mid-stream.
    }
  };
};

const dequantizeU16 = (q, minV, maxV) => minV + (q / 65535.0) * (maxV - minV);

export const decodePayloadU16 = (payload, { minV, maxV }) => {
  const bytes = toUint8Array(payload);
  if (bytes.length % 2 !== 0) {
    throw new Error('payload length must be even (uint16 packed)');
  }
  const count = bytes.length / 2;
  const values = new Array(count);
  for (let i = 0; i < count; i += 1) {
    const lo = bytes[i * 2];
    const hi = bytes[i * 2 + 1];
    const q = lo | (hi << 8);
    values[i] = dequantizeU16(q, minV, maxV);
  }
  return values;
};

export const decodeFrameU16 = (
  payload,
  { minV, maxV, rows = 12, cols = 8 }
) => {
  const bytes = toUint8Array(payload);
  const expected = 4 + rows * cols * 2;
  if (bytes.length !== expected) {
    throw new Error(`expected ${expected} bytes, got ${bytes.length}`);
  }
  const magic = bytes[0] | (bytes[1] << 8);
  if (magic !== 0xbeef) {
    throw new Error(`bad magic: ${magic.toString(16)}`);
  }
  const frameId = bytes[2] | (bytes[3] << 8);
  const values = decodePayloadU16(bytes.slice(4), { minV, maxV });
  const matrix = [];
  for (let r = 0; r < rows; r += 1) {
    matrix.push(values.slice(r * cols, (r + 1) * cols));
  }

  const rescaledMatrix = rescaleMatrix(matrix);
  return { frameId, matrix: rescaledMatrix }; 
};


export const rescaleMatrix = (matrix) => {
  const inMin = 700;
  const inMax = 3700;
  const outMin = 0;
  const outMax = 100;
  const inRange = inMax - inMin;
  const outRange = outMax - outMin;

  return matrix.map(row => row.map(value => {
    if (value <= 0) return -1;
    const scaled = ((value - inMin) / inRange) * outRange + outMin;
    return Math.min(outMax, Math.max(outMin, scaled));
  }));
};

